### PR TITLE
[CL] Serialize Incentive Record, Implement Fee Accum and Incentive Record to Genesis

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #4592

## What is the purpose of the change

Adds Fee Accumulators and incentive record to genesis. 
Adds serialization to incentive Record as a process for this.
Misc changes also included as part of refactoring.

## Brief Changelog

- Serialize Incentive Record, Implement Fee Accum and Incentive Record to Genesis

## Testing and Verifying

Tests have been added

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: proto/osmosis/concentrated-liquidity/incentive_record.proto
Comment (AlpinYukseloglu) [inline proto/osmosis/concentrated-liquidity/incentive_record.proto:15]: Just to clarify – we're generating the object w/ proto but still only storing `IncentiveRecordBody` in state right?

Comment (mattverse) [reply proto/osmosis/concentrated-liquidity/incentive_record.proto:15]: yes, there is no change in logic in the storing part, this proto struct is only used when serializing data for genesis

File: x/concentrated-liquidity/genesis.go
Comment (p0mvn) [inline x/concentrated-liquidity/genesis.go:111]: Why don't we store incentive records directly in `poolData` if we get them by pool id?

The reason we have positions separately is because we had to no way to get them from store by pool id. That can be changed now with the position key refactor

File: x/concentrated-liquidity/genesis_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/genesis_test.go:88]: Please update godoc that now position id, incentive records and accumulators are also added

File: x/concentrated-liquidity/genesis_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/genesis_test.go:289]: What about incentives accumulators? Are they stored together with fee accumulators? Can we have a test where there are incentives accumulators as well

File: x/concentrated-liquidity/genesis.go
Comment (p0mvn) [inline x/concentrated-liquidity/genesis.go:84]: Not seeing where we export incentives accumulators.

Can you please add them as well to export / import genesis?

I think we can get them using this method:
https://github.com/osmosis-labs/osmosis/blob/3aa3dffae2f6e12ee931c8891d7f90cd49b97dc8/x/concentrated-liquidity/incentives.go#L54

**Review by p0mvn — COMMENTED** [review]
Nice work.

The only blocking comment from my side is having incentives accumulators exported too

File: go.sum
Comment (AlpinYukseloglu) [inline go.sum:938]: Let's make sure to resolve this before merging

File: x/concentrated-liquidity/types/genesis/genesis.go
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/types/genesis/genesis.go:10]: Out of curiosity, what's the reason we deference this?

File: osmoutils/accum/accum.go
Comment (AlpinYukseloglu) [inline osmoutils/accum/accum.go:51]: We should ideally have tests for this to ensure it works as intended (e.g. that it actually makes the accum at `store/accum/{accumName}`) – even just a couple sanity tests would be great

**Review by AlpinYukseloglu — APPROVED** [review]
Mostly LGTM pending a couple minor comments (want to highlight the conflict in `go.sum` and @p0mvn's comments

File: go.sum
Comment (AlpinYukseloglu) [inline go.sum:944]:
```suggestion
```

File: go.sum
Comment (AlpinYukseloglu) [inline go.sum:938]:
```suggestion
```

**Review by AlpinYukseloglu — CHANGES_REQUESTED** [review]

File: x/concentrated-liquidity/types/genesis/genesis.go
Comment (mattverse) [reply x/concentrated-liquidity/types/genesis/genesis.go:10]: We want to refrain from using pointers if possible for clarity and better readability and safetyness

File: x/concentrated-liquidity/genesis_test.go
Comment (mattverse) [reply x/concentrated-liquidity/genesis_test.go:88]: Godoc updated :)

File: x/concentrated-liquidity/genesis.go
Comment (mattverse) [reply x/concentrated-liquidity/genesis.go:111]: makes sense, moved incentive records field into `PoolData`

File: x/concentrated-liquidity/genesis.go
Comment (mattverse) [reply x/concentrated-liquidity/genesis.go:84]: Gotcha, didnt know uptime accumulator == incentives accumulator, added along with necessary test cases

File: x/concentrated-liquidity/genesis_test.go
Comment (mattverse) [reply x/concentrated-liquidity/genesis_test.go:289]: ditto below

File: go.sum
Comment (mattverse) [reply go.sum:944]: Good catch! Thanks :)

File: osmoutils/accum/accum.go
Comment (mattverse) [reply osmoutils/accum/accum.go:51]: Good point! Added test cases for this

**Review by p0mvn — APPROVED** [review]
LGTM

File: osmoutils/accum/accum.go
Comment (p0mvn) [inline osmoutils/accum/accum.go:55]:
```suggestion
```

**Comment by p0mvn** [comment]
I think `osmoutils` go.mod needs to be updated for CI to pass

File: proto/osmosis/concentrated-liquidity/genesis.proto
Comment (mattverse) [reply proto/osmosis/concentrated-liquidity/genesis.proto:63]: Updated to use this proto struct instead

**Pull request merged**
